In [1]:
import numpy as np

# NumPy Interview Practice

## Part 1: Your Topics






---

## Part 2: Essential Missing Topics



### Viva Questions
1. What are the rules of Broadcasting?
2. Does `axis=0` move across rows or down columns?
3. What is Vectorization?
4. Deep copy vs View (slicing)?

### Written Questions

1. **Reshaping:** Given `arr` of shape (2, 4), convert it into a column vector (8, 1) without hardcoding rows.

In [8]:
a = np.random.randint(2,4,size=(2,4))
a.reshape(8,1)

array([[2],
       [3],
       [2],
       [2],
       [3],
       [3],
       [2],
       [2]])

2. **Flattening:** Verify code to show that modifying a `.ravel()` array affects the original, but `.flatten()` does not.

In [19]:
# b = a.flatten()
# b[2]=78
# print(b)
# print(a,"no change in original array")

c= a.ravel()
print(c)
c[3]=89
print(a,"change in original array")

[2 3 2 2 3 3 2 2]
[[ 2  3  2 89]
 [ 3  3  2  2]] change in original array


3. **Creation:** Create a 3x3 Identity matrix in one line.

In [20]:
np.eye(3,3)

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

4. **Filtering:** Use `np.where` on `[10, 20, 30, 40, 50]` to replace values > 30 with 0.

In [23]:
a = np.array([10,20,30,40,50])
np.where(a==30,0,a)

array([10, 20,  0, 40, 50])

5. **Linspace:** Generate 5 equally spaced numbers between 0 and 10.

In [25]:
np.linspace(0,10,5)

array([ 0. ,  2.5,  5. ,  7.5, 10. ])

6. **Splitting:** Split an array of 10 elements into 3 sub-arrays.

In [30]:
a = np.array([10,20,30,40,50,60,70])
np.array_split(a,3)

[array([10, 20, 30]), array([40, 50]), array([60, 70])]

7. **Random:** Generate a random integer between 50 and 100.

In [31]:
np.random.randint(50,100)

82

**1. Advanced Reshaping:**
Create an array of shape `(2, 3, 4)`. Write code to reshape it into `(6, 4)` and then transpose it to `(4, 6)` in a single line of chained commands.

In [41]:
np.random.randint(1,10,size=(2,3,4)).reshape(6,4).T

array([[5, 7, 2, 9, 8, 6],
       [9, 7, 1, 1, 1, 1],
       [9, 2, 7, 8, 7, 4],
       [4, 9, 3, 9, 6, 5]])

**2. Concatenation vs. Stacking:**
Given `a = np.array([1, 2])` and `b = np.array([3, 4])`:
* Write code to stack them **vertically** to get `[[1, 2], [3, 4]]`.
* Write code to stack them **horizontally** to get `[1, 2, 3, 4]`.

In [57]:
a = np.array([1,2])
b = np.array([3,4])

print(np.vstack([a,b]))
print(np.hstack([a,b]))

[[1 2]
 [3 4]]
[1 2 3 4]


**3. Sorting with Indices:**
You have an array of student scores: `scores = np.array([55, 89, 76, 32])`.
Write code using `np.argsort()` to print the **indices** that would sort this array. Use these indices to print the scores in ascending order.

**4. Complex Filtering (np.where):**
Given `arr = np.array([1, 5, 10, 15, 20])`:
Write a `np.where` command that replaces values < 5 with 0 AND values > 15 with 100, leaving the middle values as they are. (Hint: Nested `np.where`).

In [73]:
arr = np.array([1,5,10,15,20])

a = np.where(arr < 5,0 ,np.where(arr > 15,100,arr))
a

array([  0,   5,  10,  15, 100])

**5. Splitting Uneven Arrays:**
Create an array of 11 elements. Attempt to use `np.split()` to divide it into 3 equal parts. When it fails, write the correct code using `np.array_split()` and comment on the size of the resulting sub-arrays.

In [81]:
# b = np.random.randint(1,10,11)
print(b)

# print(np.split(b,3))

print(np.array_split(b,3))

[5 5 2 6 1 6 7 8 8 5 7]
[array([5, 5, 2, 6]), array([1, 6, 7, 8]), array([8, 5, 7])]


**6. Reproducible Randomness:**
Write code to generate a random 3x3 matrix. Then, set a "seed" using `np.random.seed()` (or `default_rng`) so that running the code again produces the **exact same** numbers.

In [91]:
# np.random.randint(1,10,9).reshape(3,3)
np.ones([3,3])

array([[1., 1., 1.],
       [1., 1., 1.],
       [1., 1., 1.]])

1. **Broadcasting:** Shape result of `(4, 1)` array + `(3,)` array?

In [98]:
a = np.random.randint(1,10,4).reshape(4,1)
b = np.random.randint(1,10,3).reshape(3,)

result = a + b

print(result)
result.shape

[[ 9  5 10]
 [12  8 13]
 [15 11 16]
 [14 10 15]]


(4, 3)

2. **Boolean Indexing:** Select even numbers from `[1, 2, 3, 4, 5]` without loops.

In [105]:
a = np.array([1, 2, 3, 4, 5])

v = a[a%2 == 0]
v

array([2, 4])

3. **Argmax:** Does `np.argmax` return the index or the value?

4. **Axis:** Calculate sum of columns in a 2D matrix.

### Viva Questions

1. Diff between `.flatten()` and `.ravel()`?

2. What does `-1` mean in `reshape`?

3. Why use `np.nditer` over loops?

4. When to use `np.empty()`?

5. How to iterate with index (`ndenumerate`)?

6. Is `np.save` format human-readable?

# 📚 Extended NumPy Practice Set

## PART 1: Deep Dive into Your Topics
*Topics covered: reshape, split, random, sort, concatenate, T, where, save/load*

### 📝 Written Test (Coding Challenges)


---

### 🗣️ Viva / Interview Questions (Conceptual)

**1. Copy vs. In-Place Sort:**
"If I run `arr.sort()`, does it return a new array or modify the existing one? How is this different from `np.sort(arr)`?"

**2. Transpose Semantics:**
"Does `arr.T` create a new copy of the data in memory, or is it just a different 'view' of the same data? Why does this matter for large datasets?"

**3. Concatenation Efficiency:**
"If I need to build a large array by adding one row at a time inside a `for` loop, should I use `np.concatenate` inside the loop? Why is this generally considered bad practice in NumPy?"

**4. Save/Load Formats:**
"You mentioned `save` and `load`. If I need to share data with a colleague who uses MATLAB or Excel, can I use `.npy` files? If not, what NumPy function should I use instead?"

---

## PART 2: Essential Missing Topics (Crucial for Interviews)
*Topics covered: Broadcasting, Boolean Masking, Aggregations, Axis Operations*

### 📝 Written Test (Coding Challenges)

**1. Broadcasting (The "Center the Data" Problem):**
Given a matrix `X` of shape `(5, 3)` (5 samples, 3 features).
Write code to subtract the **mean of each column** from the corresponding column elements. (Hint: `X - X.mean(axis=0)`).

**2. Boolean Masking (Not just `where`):**
Given `arr = np.array([10, 20, 30, 40])` and a mask `mask = np.array([True, False, True, False])`.
Write code to print only the values where the mask is True.
*Bonus:* Write code to set all values where the mask is True to -1.

**3. Dot Product:**
Create two matrices: `A` (2x3) and `B` (3x2). Calculate their matrix product (Dot Product) using both the `np.dot()` function and the `@` operator.

**4. Unique Values:**
Given an array with duplicates: `data = np.array([1, 2, 2, 3, 3, 3, 4])`.
Write code to find the unique values and their counts (frequency).

---

### 🗣️ Viva / Interview Questions (Conceptual)

**1. The "Axis" Confusion:**
"In a 2D matrix, if I ask for `np.mean(arr, axis=1)`, what does the result represent conceptually? (Row averages or Column averages?)"

**2. Vectorization vs. Loops:**
"I have a Python list of 1 million numbers. I want to square every number. Why is converting it to a NumPy array and doing `arr ** 2` hundreds of times faster than a list comprehension?"

**3. Data Types (dtype):**
"If I have a NumPy array of integers `[1, 2, 3]` and I try to insert a float `2.5` into index 0, what happens? Does it throw an error or cast the value?"

**4. Nan (Not a Number):**
"If my array contains `np.nan`, what will `np.sum(arr)` return? How do I calculate the sum while ignoring the NaNs?"

# 🥋 Ultimate NumPy Practice Set (30 Questions)

## 📌 Section 1: Array Creation & Basics
*Topics: array, dtype, arange, linspace, ones, zeros, eye, empty*

### 📝 Written Questions
1. **Type Casting:** Create an array of floats `[1.1, 2.2, 3.3]`. Write code to force convert this entire array into integers using `.astype()`.
2. **Step Size:** Use `np.arange()` to create an array of all odd numbers between 1 and 20.
3. **Precise Spacing:** Create an array of 50 values perfectly spaced between 0 and 100 using `np.linspace`.
4. **Identity:** Create a 5x5 Identity matrix using `np.eye()`.
5. **Ones & Zeros:** Create a matrix of shape `(3, 4)` filled entirely with the number `7` (Hint: Use `np.ones` * 7 or `np.full`).

### 🗣️ Viva Questions
6. **Arange vs Linspace:** "If I need exactly 9 data points between 0 and 1, which function is better to use: `arange` or `linspace`? Why?"
7. **Empty vs Zeros:** "Why is `np.empty()` slightly faster than `np.zeros()`? What kind of 'garbage' values might I find in an empty array?"
8. **Data Types:** "What is the default data type of an array created with `np.zeros((3,3))`? Is it integer or float?"

---

## 📌 Section 2: Shape Manipulation & Iteration
*Topics: shape, ndim, reshape, flatten, ravel, T, nditer, ndenumerate*

### 📝 Written Questions
9. **Dimension Check:** Given a 3D array `arr = np.ones((2, 3, 4))`, write code to print the number of dimensions (`ndim`) and the total number of elements (`size`).
10. **Reshape Logic:** Create an array of 24 elements (`arange(24)`). Reshape it into dimensions `(2, 3, 4)`.
11. **Negative Indexing in Reshape:** Reshape the previous array into shape `(6, ?)`, letting NumPy figure out the second dimension automatically.
12. **Iteration:** Write a loop using `np.ndenumerate` that prints the index coordinate (e.g., `(0, 1)`) alongside the value for every element in a 2x2 matrix.

### 🗣️ Viva Questions
13. **Transpose:** "Does `arr.T` physically move data in memory? Explain the concept of 'strides'."
14. **Ravel vs Flatten:** "I have a 1GB array. If I use `ravel()`, does memory usage double? What if I use `flatten()`?"

---

## 📌 Section 3: Indexing, Slicing & Logic
*Topics: where, split, array_split, boolean indexing*

### 📝 Written Questions
15. **Conditional Replacement:** Given `arr = np.array([1, 2, 3, 4, 5])`, use `np.where` to print `"Even"` for even numbers and `"Odd"` for odd numbers.
16. **Boolean Masking:** Given `arr = np.array([10, 20, 30, 40])`, write code to select only the values **greater than 25**.
17. **Splitting:** Split an array of 20 elements into 4 equal-sized sub-arrays.
18. **Uneven Split:** Try to split an array of 10 elements into 3 parts using `np.array_split`. Print the lengths of the resulting chunks.

### 🗣️ Viva Questions
19. **Where Logic:** "Can `np.where` be used to merge data from two different arrays based on a condition? Explain how."
20. **Slicing vs Copying:** "If I do `b = a[::2]`, is `b` a copy of `a`? If I modify `b`, does `a` change?"

---

## 📌 Section 4: Math, Stats & Broadcasting (The "Missing" Essentials)
*Topics: add, sum, mean, axis, broadcasting*

### 📝 Written Questions
21. **Column Sums:** Generate a 4x4 matrix of random numbers. Calculate the sum of each **column**.
22. **Row Averages:** Using the same matrix, calculate the **mean** of each **row**.
23. **Broadcasting Basics:** specific Create an array `A` of shape `(3, 3)` and add a 1D array `b` of shape `(3,)` to it. Print the result.
24. **Global Stats:** Find the maximum, minimum, and standard deviation of a random 100-element array.

### 🗣️ Viva Questions
25. **Axis Confusion:** "Explain `axis=0` vs `axis=1` using the analogy of 'collapsing' dimensions."
26. **Broadcasting Rule:** "Can I add an array of shape `(2, 5)` to an array of shape `(5,)`? Why or why not?"

---

## 📌 Section 5: Random, Sorting & I/O
*Topics: random.rand, randint, sort, argsort, save, load*

### 📝 Written Questions
27. **Random Integers:** Generate a 3x3 matrix of random integers between 1 and 100.
28. **Arg-Sorting:** Given `arr = [50, 10, 40, 30]`. Use `np.argsort()` to find the index of the smallest element.
29. **Save & Load:** Write code to save a random array to a file named `my_data.npy` and then load it back into a variable.

### 🗣️ Viva Questions
30. **Random vs Normal:** "What is the difference between `np.random.rand()` and `np.random.randn()`?"
31. **Sorting Dimensions:** "If I use `np.sort(arr, axis=0)` on a 2D matrix, does it sort the rows independently or the columns independently?"